contracts/andromeda_crowdfund/src/contract.rs

```diff
@@ -0,0 +1,563 @@
+        deps.storage,
+        deps.querier,
+        sender.clone(),
+        Funds::Native(state.price.clone()),
+        encode_binary(&"")?,
+    )?;
+    let remaining_amount = remainder.try_get_coin()?;
+
+    state.amount_to_send += remaining_amount.amount;
+
+    let tax_amount = get_tax_amount(&msgs, state.price.amount, remaining_amount.amount);
+    // require that the sender has sent enough for taxes
+    require(
+        has_coins(
+            &info.funds,
+            &Coin {
+                denom: state.price.denom.clone(),
+                amount: state.price.amount + tax_amount,
+            },
+        ),
+        ContractError::InsufficientFunds {},
+    )?;
+
+    UNAVAILABLE_TOKENS.save(deps.storage, &token_id, &false)?;
+
+    let purchase = Purchase {
+        token_id: token_id.clone(),
+        tax_amount,
+        msgs,
+        purchaser: sender.clone(),
+    };
+
+    purchases.push(purchase);
+    PURCHASES.save(deps.storage, &sender, &purchases)?;
+
+    state.amount_sold += Uint128::from(1u128);
+    STATE.save(deps.storage, &state)?;
+
+    Ok(Response::new()
+        .add_attribute("action", "purchase")
+        .add_attribute("token_id", token_id))
+}
+
+fn execute_claim_refund(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+) -> Result<Response, ContractError> {
+    let state = STATE.may_load(deps.storage)?;
+    require(state.is_some(), ContractError::NoOngoingSale {})?;
+    let state = state.unwrap();
+    require(
+        state.expiration.is_expired(&env.block),
+        ContractError::SaleNotEnded {},
+    )?;
+    require(
+        state.amount_sold < state.min_tokens_sold,
+        ContractError::MinSalesExceeded {},
+    )?;
+
+    let purchases = PURCHASES.may_load(deps.storage, info.sender.as_str())?;
+    require(purchases.is_some(), ContractError::NoPurchases {})?;
+    let purchases = purchases.unwrap();
+    let refund_msg = process_refund(deps.storage, &purchases, &state.price);
+    let mut resp = Response::new();
+    if let Some(refund_msg) = refund_msg {
+        resp = resp.add_message(refund_msg);
+    }
+
+    Ok(resp.add_attribute("action", "claim_refund"))
+}
+
+fn execute_end_sale(
+    deps: DepsMut,
+    env: Env,
+    limit: Option<u32>,
+) -> Result<Response, ContractError> {
+    let state = STATE.may_load(deps.storage)?;
+    require(state.is_some(), ContractError::NoOngoingSale {})?;
+    let state = state.unwrap();
+    require(
+        state.expiration.is_expired(&env.block),
+        ContractError::SaleNotEnded {},
+    )?;
+    if state.amount_sold < state.min_tokens_sold {
+        issue_refunds_and_burn_tokens(deps, env, limit)
+    } else {
+        transfer_tokens_and_send_funds(deps, env, limit)
+    }
+}
+
+fn issue_refunds_and_burn_tokens(
+    deps: DepsMut,
+    env: Env,
+    limit: Option<u32>,
+) -> Result<Response, ContractError> {
+    let state = STATE.load(deps.storage)?;
+    let limit = limit.unwrap_or(DEFAULT_LIMIT) as usize;
```

Comment:
Does this need a non-zero check?

Reply:
I think it is fine without since if they put zero it will just not process anything. For the other function I added the zero check to avoid the weird case it would pose with taking `limit + 1`. It might also be fine there but just to be safe.